src/pymor/algorithms/rand_la.py

```diff
@@ -0,0 +1,350 @@
+        Q = source_product.apply_inverse(Q)
+        gram_schmidt(Q, source_product, atol=0, rtol=0, copy=False)
+        Q = A.apply(Q)
+        gram_schmidt(Q, range_product, atol=0, rtol=0, copy=False)
+
+    if return_rand:
+        return Q, R
+    else:
+        return Q
+
+
+    @defaults('p', 'q')
+    def random_generalized_svd(A, S=None, T=None, modes=None, p=20, q=2):
+    
+    """ Randomized SVD of a |VectorArray|. 
+    
+    Viewing the |VectorArray| 'A' as a 'A.dim' x 'len(A)' matrix, the return value
+    of this method is the randomized singular value decomposition of 'A', where the
+    inner product on R^('dim(A)') is given by 'S' and the inner product on R^('len(A)')
+    is given by 'T'. 
+        
+        A = U*\Sigma*V*T
+    
+    Parameters
+    ----------
+    A : 
+        The |VectorArray| or |Operator| for which the randomized SVD is to be computed. 
+    S : 
+        Range product |Operator| w.r.t which the randomized SVD is computed
+
+    T :
+        Source product |Operator| w.r.t which the randomized SVD is computed
+
+    modes : 
+        If not 'None', at most the first 'modes' approximated singular values 
+        and vectors are returned. 
+    p :
+        If not '0', adds 'Oversampling' colums to the randomly sampled matrix 'G'. 
+    q : 
+        If not '0', performs 'PowerIterations' iterations to increase the relative weight
+        of the first singular values.
+
+    Returns
+    -------
+    U 
+        |VectorArray| of approximated left singular vectors 
+    s 
+        One-dimensional |NumPy array| of the approximated singular values
+    V
+        |VectorArray| of the approximated right singular vectors
+    """
+    logger = getLogger('pymor.algorithms.rand_la')
+    
+    assert isinstance(A, VectorArray) or isinstance(A, Operator)
+    
+    if isinstance(A, VectorArray):
+        A = VectorArrayOperator(A)
+   
+    assert 0 <= modes <= max(A.source.dim, A.range.dim) and isinstance(modes,int)
+    assert p >= 0 and isinstance(p, int)
+    assert q >= 0 and isinstance(q, int)
+    
+    if S is None:
+        S = IdentityOperator(A.range)
+    else:
+        assert isinstance(S, Operator)
+        
+    if T is None:
+        T = IdentityOperator(A.source)
+    else:
+        assert isinstance(T, Operator)
+        
+    if A.source.dim == 0 or A.range.dim == 0:
+        return A.source.empty(), np.array([]), A.range.empty()
+    
+    Q = rrf(A, source_product=T, range_product=S, q=q, l=modes+p)
+    B = A.apply_adjoint(S.apply(Q))
+    Q_B, R_B = gram_schmidt(T.apply_inverse(B), product=T, return_R=True)
+    U_b, s, V_b = sp.linalg.svd(R_B.T)
+
+    with logger.block(f'Computing generalized left-singular vectors ({U_b[:,:modes].shape[1]} vectors) ...'):
+        U = Q.lincomb(U_b)
+
+    with logger.block(f'Computung gerneralized right-singular vector ({V_b[:modes,:].shape[0]} vectors) ...'):
+        V = Q_B.lincomb(V_b.T)
+        
+    return  U[:modes], s[:modes], V[:modes]
```

Comment:
Even though the docstring says otherwise, this functions works with arbitrary `Operators`. Hence, in contrast to the `VectorArray` svd methods where the source of the corresponding `Operator` is 'special', it's actually not possible/feasible to return a numpy array here in general.

I admit having both SVD version in pyMOR (for `Operators` and `VectorArrays`) now things seem to be a bit confusing. The two options I see are:
- Add proxy methods for randomized SVDs of `VectorArrays` to `svd_va` which wrap the array into a `VectorArrayOperator` (remove the option of passing `VectorArrays` here) and then call `V.to_numpy` in the proxy method.
- Return `Vh` in `svd_va` as `NumpyVectorArray` such that everything has the same return type.

I have a slight preference towards the first option. Other opinions, @pymor/all?

Reply:
No preference for either.

Reply:
I'm thinking that it might also be convenient to be able to perform a regular SVD of an operator. Would it be possible to unify the approaches by modifying the `svd_va` module to also accept operator inputs?

Reply:
The second option sounds good to me. I am against the first option, for a similar reason that there exists separate `svd` and `svds` methods in SciPy (methods in `svd_va` are similar to SciPy's `svd`, while randomized SVD is more like `svds`). It might also be good to rename this method to `randomized_svds` or similar. We could think about having an `svds` method based on `eigs`.

Having a method for regular SVD of dense matrices might actually be good, because then the SVD solver could be chosen through `defaults`. I generally used `gesvd` for the accuracy, but SciPy recently got a wrapper for `gejsv`, which should be even more accurate (of course, there is also the faster but less accurate `gesdd`). Similar to dense matrix equation solver methods, I would be for such an SVD method to only accept NumPy arrays and not `Operators`, to not give wrong expectations to the user. We could rename `svd_va` to just `svd` and add the method for NumPy arrays there.

Reply:
>  I am against the first option, for a similar reason that there exists separate svd and svds methods in SciPy (methods in svd_va are similar to SciPy's svd, while randomized SVD is more like svds)


I'm not completely sure I can follow your argument. For me this is not a question about dense vs. sparse methods but about the data types these algorithms operate on.

> The weak point here is probably that this would not work for arbitrary operators.

Yes, conversion of `Operators` into `VectorArrays` is only possible under very special circumstances and, IMHO, should always require explicit conversion by the user except in places where it is known that the conversion is safe. (E.g. calling `B.as_range_array()` for an `LTISystem`.)

Probably some more discussion on how to organize things is needed. I do not think it is necessary to block this PR for this, so I suggest to leave `svd_va` as is and to actually remove the implicit `VectorArray` to `Operator` conversion in the randomized methods. Everything else can be handled in a separate PR. I will put the issue on the agenda for the next dev meeting.

Reply:
> > I am against the first option, for a similar reason that there exists separate svd and svds methods in SciPy (methods in svd_va are similar to SciPy's svd, while randomized SVD is more like svds)
> 
> I'm not completely sure I can follow your argument. For me this is not a question about dense vs. sparse methods but about the data types these algorithms operate on.

I would say that these are related issues, i.e., I think `svd` and `svds` are different enough such that they could operate on different data types. Also, I don't see what would be the benefit of applying randomized SVD to a `VectorArray` compared to using one of the methods in `svd_va`.
Maybe we should again also discuss how to handle low-rank and dense matrix equation solvers?

Reply:
> I would say that these are related issues, i.e., I think svd and svds are different enough such that they could operate on different data types.

I am really a bit lost. Can you define what exactly mean by `svd` and `svds`. I assume that historically the `s` stands for sparse, but I assume you are thinking of truncated vs. non-truncated svd? 

> Also, I don't see what would be the benefit of applying randomized SVD to a VectorArray compared to using one of the methods in svd_va.

Better performance when you only need a few leading singular values?

> Maybe we should again also discuss how to handle low-rank and dense matrix equation solvers?

Probably ..

Reply:
> I am really a bit lost. Can you define what exactly mean by `svd` and `svds`. I assume that historically the `s` stands for sparse, but I assume you are thinking of truncated vs. non-truncated svd?

Now that I think about it, in essence, `svd` is an SVD method that needs access to the matrix entries, while `svds` uses only `apply` and `apply_adjoint`. Typically, `svds` would be used to only compute a few largest (or smallest) singular values and corresponding singular vectors, but I guess that technically it could be used to compute the full SVD. On the other hand, `svd` computes the full SVD that can be truncated.

In that sense, SVD methods using the `VectorArray` interface would be a third class of methods.

> > Also, I don't see what would be the benefit of applying randomized SVD to a VectorArray compared to using one of the methods in svd_va.
> 
> Better performance when you only need a few leading singular values?

I'm wondering if randomized SVD is still justified when `source.dim` or `range.dim` is small.

Reply:
> I am really a bit lost. Can you define what exactly mean by svd and svds. I assume that historically the s stands for sparse, but I assume you are thinking of truncated vs. non-truncated svd?

I hope I don't explain the obvious here but I think in Scipy (and Matlab) `svd` refers to a 'conventional' SVD algorithm via QR, whereas `svds` uses some kind of Lanczos Bidiagonalization algorithm of [0 A ; A^T 0] which can be seen as a sparse matrix. Usually, a apply methods for A and A^T are supplied to `svds` instead of the matrix A.

> The second option sounds good to me. I am against the first option, for a similar reason that there exists separate svd and svds methods in SciPy (methods in svd_va are similar to SciPy's svd, while randomized SVD is more like svds).

I agree that randomized SVD is similar to `svds` in the sense that it can work with `apply` methods, but after the projections, usually a regular QR-based `svd` is performed so it is different in that regard.

> I'm wondering if randomized SVD is still justified when source.dim or range.dim is small.

For me, this is what it comes down to.

Do you guys think it would be possible to have an SVD wrapper that takes Operators as well as VectorArrays and decides on the correct method based on the input type, dimensions and tolerances considering error and runtime estimates?

Reply:
@pmli, I think this entire discussion shows that `svds` would not be a good name for 'low-rank svd' in pyMOR as people might expect a specific algorithm. Since all svds in pyMOR are truncated, I would think that a single common signature for all svd algorithms (at least for `Operators`) would be sufficient.


> >  I'm wondering if randomized SVD is still justified when source.dim or range.dim is small.

> For me, this is what it comes down to. It is probably unjustified, but it would still be nice to reap the benefits of efficient apply methods, like NumpyHankelOperator and hopefully more classes like this in the future. At the moment this is not possible but svds could be used in these unjustified cases (ideally automatically).

For tall and very skinny matricies, a randomized svd is probably not so useful. If you have a 10^8 x 10^3 matrix, I'd say that a randomized method might be very beneficial.

> Do you guys think it would be possible to have an SVD wrapper that takes Operators as well as VectorArrays and decides on the correct method based on the input type, dimensions and tolerances considering error and runtime estimates?

So I am not convinced that it is so useful to have a single wrapper which takes both `Operators` and `VectorArrays`. However, we are discussing of having a unified approach for wrappers for svd and other numerical linear algebra problems (linear problems, matrix equations) that dispatch to specific algorithms based on data type / heuristics or user choice. However, we should not have that discussion in this pr ..

Reply:
> So I am not convinced that it is so useful to have a single wrapper which takes both Operators and VectorArrays. However, we are discussing of having a unified approach for wrappers for svd and other numerical linear algebra problems (linear problems, matrix equations) that dispatch to specific algorithms based on data type / heuristics or user choice. However, we should not have that discussion in this pr ..

Yes, I will keep quiet. I also feel like I am on a different page to you guys :woozy_face:
It would be nice, if you could put me in the loop regarding this topic, i.e. let me know where the discussion moves to or inform me about any results. I don't think I will have much to contribute, but I think this will influence my work :)

Reply:
> It would be nice, if you could put me in the loop regarding this topic

Will do.